sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/metrics/TimeMonitor.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.loadtests.metrics;
+
+import org.apache.beam.sdk.metrics.Distribution;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+
+/** Monitor that records processing time distribution in the pipeline. */
```

Comment:
It seems there's an intended 'recipe' for how to use this utility; could you either document it? i.e.

```java
/**
 *  Monitor that records processing time distribution in the pipeline.
 *
 * To use: apply a monitor directly after each source and sink transform.
 * This will capture a distribution if element processing timestamps, which can be
 * collected an written out using {@link MetricsPublisher}.
```

Another idea (and I don't know how feasible this is) would be to add an API to automatically apply the intended recipe to a pipeline definition. Something like `TimeMonitor.injectInto(pipeline, "namespace", "counterName")` This might be possible using [`Pipeline.replaceAll()`](https://github.com/apache/beam/blob/658630d8f75281f50dc434f2e062e2819ebeff84/sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java#L206) or similar.

Reply:
I investigated this a bit. 

For start time collection it looks simple. We could create a pipeline factory method that returns it with a "substitute" of source, like so: 

```
public PCollection<KV<byte[], byte[]>> getPipelineWithAppliedSource() {
    pipeline = Pipeline.create(options);
    return pipeline.apply(Create.of(KV.of(new byte[0], new byte[0])));

  }
```
after that we could replace this easily with a composite transform of source + `ParDo.of(TimeMonitor)`


A question though:
 - the code you mentioned is in _beam-runners-core-construction-java_. That means the _load_tests_ package would need to depend on it. Is it ok to have such dependency? 

And some concerns: 
 - I didn't check well enough yet but it seems feasible to look for all the leaves of the Job DAG, am I right (or maybe I didn't see something coming)? That would be needed to find out where should we put TimeMonitors to collect end times. 
 -  If the type of a job graph is different than `PCollection<KV<byte[], byte[]>>` then we might have trouble because I don't think we can infer what types TimeMonitor should accept. 

Let me know what you think. Either way, I suggest creating a separate ticket/PR for this too in order not to block other work.

Reply:
> the code you mentioned is in beam-runners-core-construction-java. That means the load_tests package would need to depend on it. Is it ok to have such dependency?

[`Pipeline.replaceAll()`](https://github.com/apache/beam/blob/658630d8f75281f50dc434f2e062e2819ebeff84/sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java#L206) is part of `beam-sdks-java-core`, so it would be fine to depend on.

> I didn't check well enough yet but it seems feasible to look for all the leaves of the Job DAG, am I right (or maybe I didn't see something coming)? That would be needed to find out where should we put TimeMonitors to collect end times.

I worked with this code myself, so I can't attest to the pitfalls we might find. But the [`PTransformMatcher`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformMatcher.java) matches on an [`AppliedPTransform`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/runners/AppliedPTransform.java), which has `getInputs()` and `getOutputs()` APIs. Then you would implement a [`PTransformOverrideFactory`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformOverrideFactory.java) `getReplacementTransform()` and then somehow wrap each source/sink to include the monitor.

However, an issue for sources is that they must be at a root of the DAG, so we cannot insert the monitor before. Wrapping the source in a composite to get the type signature right won't help; runners will expand composites to their primitives for execution and the source would still be a root. This would be a blocker until we have Splittable DoFns which can receive input.

> If the type of a job graph is different than PCollection<KV<byte[], byte[]>> then we might have trouble because I don't think we can infer what types TimeMonitor should accept.

That's a good point. To be more general we could possibly try to use the type coder and record encoded byte size. But, this is getting complicated..

> Let me know what you think. Either way, I suggest creating a separate ticket/PR for this too in order not to block other work.

The more we talk about it, the more it sounds like overkill. I would recommend discarding this idea as impractical.

Reply:
> Pipeline.replaceAll() is part of beam-sdks-java-core, so it would be fine to depend on.

My mistake, you're right.

> The more we talk about it, the more it sounds like overkill. I would recommend discarding this idea as impractical.

Now I think the same. Thanks for the answer anyway - I learned a lot!